### ISSUE TITLE
Filter changesets in History by bounding box area

### ISSUE THREAD

**Issue by lectrician1** [openstreetmap/openstreetmap-website#3242]
A feature such as a slider that can filter the changesets by their bounding box area could help mappers sort through changesets in the area they are looking at without seeing the annoying larger changesets that cover the entire globe.

**Comment by tordans** [comment]
@lectrician1 do you know osmcha? You can setup filter which only show changes inside a specified bbox.
You need to also set "bbox size bound to 1" (see https://github.com/mapbox/osmcha-frontend/search?q=bbox+size&type=issues for more).
You can setup an rss feed and such as described in https://www.openstreetmap.org/user/wille/diary/47903.

**Comment by nitramwin** [comment]
@lectrician1 this would be really really great. I have just submitted an issue for iD to have a new 'check' to reduce the number of unintentional large-area edits https://github.com/openstreetmap/iD/issues/8590

**Comment by Lee-Carre** [comment]
@tordans

> osmcha? You can setup filter which only show changes inside a specified bbox.

I **have** to use OSMCha with the filtering you describe, because the [osm.org/history](http://www.openstreetmap.org/history#map=10/49.2087/-2.1235) list is flooded with so many non-local changesets as to make it unusable.

The problem, then, is that the behaviour is not as you describe.

While changesets with bounding boxes larger than my area of interest are indeed excluded, also excluded are changesets which *do* make local changes.

Besides, a tool-dev (@mmd-osm) has [said](http://www.openstreetmap.org/changeset/110409629) that
> Today there's no workable solution available to analyze large changesets.

So, even with OSMCha, I either have to manually wade through dozens of completely irrelevant changesets (which make zero local changes), or have to be unaware of larger bounding-box changesets which do make some local change(s).

While OSMCha is a tool which I generally like, it's also v0, meaning that the developers think it's not yet ready for prime-time (which would be v1+).

**Comment by mmd-osm** [comment]
For avoidance of doubt, I'm not involved in OSMCha development. The prototype I have presented uses alternative Overpass backends, whjile leveraging OSMCha's changeset display components to demo improvements.

**Comment by boothym** [comment]
Yes, a tick box for "only show changesets in view" would be very useful - so if you are looking at one town or city you can see just the ones wholly contained in your current viewport.

**Comment by mikelmaron** [comment]
> While OSMCha is a tool which I generally like, it's also v0, meaning that the developers think it's not yet ready for prime-time (which would be v1+).

I don't think that's quite accurate and depends on what's meant by "prime-time". It's very much in production shape now. Could see it being a great service to integrate into openstreetmap-website. That would need some thinking (and maybe what you mean) -- particularly assessing usage volume and making sure infrastructure is in place, or hosted on OSMF infrastructure itself.

**Comment by kallejre** [comment]
I had similar issue today where after holiday season i wanted to catch up with changes made in local area over the past few weeks, but sadly there was recently very large undiscussed import, which was promptly followed by series revertions from DWG. So, how about adding option to OSM website's history (osm.org/history) view to hide changesets that don't fit into current bounding box? Luckily there's already this issue, so i won't create a duplicate one.

From technical viewpoint, this seems to be relatively simple change, adding around 10-15 lines of code in [backend](https://github.com/openstreetmap/openstreetmap-website/blob/f1c6a87aa137c11d0aff5a4b0e563ac2c2a8f82d/app/controllers/api/changesets_controller.rb#L258-L261) for how changesets are filtered (just additional nested if-statement and support for two new parameters on ln162), 15-20 lines for [adding checkbox to frontend UI](https://github.com/openstreetmap/openstreetmap-website/blob/d02b8d9f1ed9b99651cdffae69b8c70e533d5517/app/views/changesets/history.html.erb#L15) and passing checkbox and slider value to API in [update() function](https://github.com/openstreetmap/openstreetmap-website/blob/70d7d8d850148f714b70a3297c02a8203214dec6/app/assets/javascripts/index/history.js#L62). I think I'm missing here something from what receives response from API and converts it into pretty html.

Here's quick UI prototype for filtering. Slider could disappear when checkbox is unchecked. Some users might prefer numeric entry for determining how large bounding boxes should be selected, but i felt that having to type some value into the field would feel unnatural and inconsistent with rest of website's mouse-oriented UI. For same reason checkbox and slider don't have "Submit" button, if anyone wonders.

<details>
      <summary>HTML code for the UI prototype below</summary>

```html
<div class="add-meaningful-name">
  <input type="checkbox" id="hide-cs" name="hide-cs" onchange="update()">
  <label for="hide-cs">Hide very large changesets</label><br>
  <span class="add-name-for-collapsible-class">
    Select size sensitivity:<br>
    Smaller&numsp;<input type="range" min="1" max="5" value="2" class="chset-bbox-slider" onchange="update()">&numsp;Larger
  </span>
</div>
```
</details>

![Prototype UI](https://cdn.discordapp.com/attachments/607265062322700308/928616416356237312/unknown.png)

Another portion is so-called "business decisions" for determining the actual behaviour of the slider.  I don't think that hiding a large changeset could hide many relevant changesets mostly due to probability of how changed elements are distributed around changeset area. The larger the changeset area, the less likely it is to contain modifications in currently viewed area. Obviously this should be optional feature deselected by default, but looking at large changesets, here in Estonia they are usually irrelevant changes often focusing on Russia or Norway.

Questionable part is deciding where goes the borderline for selection of the changeset. Initial idea was to show only changesets which fit fully in the viewport. Another suggestion involved calculating area of bounding box and comparing it against area of viewport, but I'm afraid performing such calculations might be too demanding for the server. Also, said area can't be a constant due to map having multiple zoom levels. Perhaps ignore changesets, where current viewport area is less than X% of the changeset area? X is determined by the slider.

Finally, we are talking here about spatial database queries done on server infra possibly millions times per second, meaning the more refined changeset filtering algorithm is selected, more significant performance impact it would have. Computationally least expensive could be following filter (based on [existing method](https://github.com/openstreetmap/openstreetmap-website/blob/f1c6a87aa137c11d0aff5a4b0e563ac2c2a8f82d/app/controllers/api/changesets_controller.rb#L258-L261)):
```ruby
changesets.where("min_lon < ? and max_lon > ? and min_lat < ? and max_lat > ?",
                         bbox.min_lon.to_i, bbox.max_lon.to_i,
                         bbox.min_lat.to_i, bbox.max_lat.to_i)
```

PS. Probably I don't know how Ruby works because i have never used it, but shouldn't API [`changesets` endpoint](https://github.com/openstreetmap/openstreetmap-website/blob/f1c6a87aa137c11d0aff5a4b0e563ac2c2a8f82d/app/controllers/api/changesets_controller.rb#L347) also support URL parameters `open/closed=false`, currently it seems to only check if parameter is present, not it's value. Therefore code from that part is probably not reusable for the changeset filtering methods.

**Comment by Lee-Carre** [comment]
@mikelmaron

> > While OSMCha is a tool which I generally like, it's also v0, meaning that the developers think it's not yet ready for prime-time (which would be v1+).
> 
> I don't think that's quite accurate and depends on what's meant by "prime-time". It's very much in production shape now.

From [*The Art of Unix Programming*](http://www.catb.org/esr/writings/taoup/) (§ Part Ⅲ. Implementation § Chapter 16. Reuse) § [The Best Things in Life Are Open](http://www.catb.org/esr/writings/taoup/html/ch16s04.html):

> …
> …. A major consequent difference between open-source practice and elsewhere is that a release level of 1.0 actually means the software is ready to use. In fact, a version number of 0.90 or above is a fairly reliable signal that the code is production-ready, but the developers are not quite ready to bet their reputations on it.
> …

Authored by Eric Raymond. The same author of *The Cathedral and the Bazaar*.

I didn't state that *I* thought it wasn't ready for non-experimental usage, but that the *developers* didn't yet think it was (by their own standards).
Re-read my original.

However, when fixes to mobile layout occurred only recently (v0.84, if I recall) … well, QED methinks.

More often, version<1 is a comment on feature-completeness, rather than reliability, robustness, refinement (debugging) & such.

> That would need some thinking […] -- particularly assessing usage volume and making sure infrastructure is in place …

See above 😄.

**Comment by mikelmaron** [comment]
@Lee-Carre I know the developers closely. I don't think use of version numbers, wherever they are cited from, is a good way to assess anything here in the OSM ecosystem -- please note the version number of the OSM API ;)

**Comment by mmd-osm** [comment]
> I don't think that's quite accurate and depends on what's meant by "prime-time". It's very much in production shape now.

As long as mappers are facing issues with OSMCha when analyzing large changesets like described in [this blog post](https://www.openstreetmap.org/user/Supaplex/diary/398447). there's probably still some room for improvement. In this case OSMCha is not to blame. It's an Overpass limitation as Supaplex correctly noted, that simply needs to get fixed.

**Comment by kallejre** [comment]
Almost 2 months without any updates. What is current status of this issue? Is it abandoned? What stops it from resolving? If it's no longer wanted, why not close the issue then?

I'd like to see some sort minimalistic feature to filter changesets on website, because unlike other 3rd party tools, OSM website offers seamless transition between 3 main activities mapper would do: viewing, editing and reviewing.

**Comment by nitram9** [comment]
I can't speak for others but I am still very much in favor of
implementation of something along the lines of what @kallejre posted on
06JAN22.

**Comment by HolgerJeromin** [comment]
> Almost 2 months without any updates. 

I am shocked! (not)
The oldest open issue in this repo is over nine years old. No need to panic. 
Most/all work here is done by unpaid developers.

> What is current status of this issue? Is it abandoned? What stops it from resolving? 

Interest and time of a developer probably.

> I'd like to see...

You can go ahead and implement it yourself! The code is right here.
